new/resources-anywhere.md

```diff
@@ -0,0 +1,211 @@
+
+## Motivation
+
+    As a <<user_profile>>,
+    I want to <<functionality>>,
+    so that <<benefit>>.
+
+## Specification
+
+### Resources outside of `Chef::Resource`
+
+Users who wish to register a resource named `my_resource` can presently do it by creating a class inside the `Chef::Resource` namespace and another in `Chef::Provider`.  Chef then uses magical namespace lookups to
+
+```ruby
+# chef/resource/
+class Chef::Resource::MyResource < Chef::Resource
+end
+```
+
+With this proposal, the above will still work; but users can now create resources outside of Chef::Resource with the following classes:
+
+```ruby
+class MyResource < Chef::Resource
+  ...
+  class Provider < Chef::Provider
+    ...
+  end
+end
+```
+
+The mechanism is:
+- An `inherited` hook on `Chef::Resource` automatically takes care of hooking up `MyResource` to all recipe DSL when the class is created.
+- The provider lookup for a resource looks first for a `Provider` class inside the Resource before moving on to other things.
+
+### Accessing `File`
+
+Accessing top-level objects like `File` are one of the big reasons for this proposal.  Anything that extends from `Chef::Resource` *will* look up `Chef::Resource::X` before `::X` when a method types `X`.
+
+To fix this, we will move resources and providers in `Chef::Provider::*` and `Chef::Resource::*` into `Chef::Resources` (and use the inline syntax as well).  For addressing the obvious concerns, see the Errors and Deprecations section.
+
+### Removing `method_missing`
+
+Presently, when you type `my_resource` in the DSL, there is no actual method by that name; we intercept `method_missing`, perform a string transformation on the method name to get `MyResource`, and then try to get the `Chef::Resource::MyResource` class.  This causes confusing stack traces when the resource cannot be looked up, and is significantly more difficult to read.
+
+In this new world, a `Chef::Resources.my_resource` method is created, and `Chef::Resources` is mixed into the recipe DSL, so there is a real method.  `method_missing` never gets involved.  The methods generated look like this:
+
+```ruby
+def my_resource(name, created_at=nil, &block)
+  created_at ||= caller[0]
+  declare_resource(:my_resource, name, created_at, ::MyResource, &block)
+end
+```
+
+NOTE: This does not remove *all* of method_missing; in particular, we need to make definitions do something similar.  Add that to the proposal.
+
+### Declaring a Provider inline (Also Single-File Resources)
+
+In order to create a Resource, you must have a corresponding Provider.  The current algorithm takes `Chef::Resource::MyResource`, lops off `Chef::Resource::` entirely, and prepends `Chef::Provider::`.  This is The new algorithm will first look for a `Chef::Resource::MyResource::Provider` class, so that declaring the provider for a class will look like this for most people:
```

Comment:
Thanks!  I'll update.  It will still be super nice for stack traces and method_missingness if we can do the wire-up once (also for performance and readability) to actual methods.